base/libdl.jl

```diff
@@ -347,34 +354,76 @@ end
+Helper type for lazily constructed library paths within the Julia distribution.
+Constructs paths relative to Julia's private shared library directory.
 
+Primarily used by Julia's standard library. For example:
+```julia
+const libgmp = LazyLibrary(BundledLazyLibraryPath("libgmp.so.10"))
 ```
-libfoo = LazyLibrary(BundledLazyLibraryPath("libfoo.so.1.2.3"))
-```
+
+!!! compat "Julia 1.11"
+    `BundledLazyLibraryPath` was added in Julia 1.11.
+
+See also [`LazyLibrary`](@ref), [`LazyLibraryPath`](@ref).
 """
 BundledLazyLibraryPath(subpath) = LazyLibraryPath(PrivateShlibdirGetter(), subpath)
 
 # Small helper struct to initialize a LazyLibrary with its initial set of dependencies
-struct InitialDependencies
-    dependencies::Vector{Any}
+struct InitialDependencies{T}
+    dependencies::Vector{T}
 end
-(init::InitialDependencies)() = convert(Vector{LazyLibrary}, init.dependencies)
+(init::InitialDependencies)() = copy(init.dependencies)
 
 """
-    LazyLibrary(name, flags = <default dlopen flags>,
+    LazyLibrary(name; flags = <default dlopen flags>,
                 dependencies = LazyLibrary[], on_load_callback = nothing)
 
-Represents a lazily-loaded library that opens itself and its dependencies on first usage
-in a `dlopen()`, `dlsym()`, or `ccall()` usage.  While this structure contains the
-ability to run arbitrary code on first load via `on_load_callback`, we caution that this
-should be used sparingly, as it is not expected that `ccall()` should result in large
-amounts of Julia code being run.  You may call `ccall()` from within the
-`on_load_callback` but only for the current library and its dependencies, and user should
-not call `wait()` on any tasks within the on load callback.
+Represents a lazily-loaded shared library that delays loading itself and its dependencies
+until first use in a `ccall()`, `@ccall`, `dlopen()`, `dlsym()`, `dlpath()`, or `cglobal()`.
+This is a thread-safe mechanism for on-demand library initialization.
+
+# Arguments
+
+- `name`: Library name (or lazy path computation) as a `String`,
+  [`LazyLibraryPath`](@ref), or [`BundledLazyLibraryPath`](@ref).
+- `flags`: Optional `dlopen` flags (default: `RTLD_LAZY | RTLD_DEEPBIND`). See [`dlopen`](@ref).
+- `dependencies`: Vector of `LazyLibrary` object references to load before this one.
+- `on_load_callback`: Optional function to run arbitrary code on first load (use sparingly,
+  as it is not expected that `ccall()` should result in large amounts of Julia code being run.
+  You may call `ccall()` from within the `on_load_callback` but only for the current library
+  and its dependencies, and user should not call `wait()` on any tasks within the on load
+  callback as they may deadlock).
+
+The dlopen operation is thread-safe: only one thread loads the library, acquired after the
+release store of the reference to each dependency from loading of each dependency. Other
+tasks block until loading completes. The handle is then cached and reused for all subsequent
```

Comment:
When you say other tasks, you mean other `dlopen()` tasks, right?  Not all tasks?